### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO) and LLVM BOLT for Perspective C++ data engine

### DISCUSSION THREAD

**Discussion by zamazan4ik** [perspective-dev/perspective#2406] [Ideas]
Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on many projects. All current results are available [here](https://github.com/zamazan4ik/awesome-pgo/). According to the multiple tests, PGO can help with improving performance in many cases (including databases). Since Perspective [cares](https://github.com/finos/perspective#features) about the performance, I think trying to optimize Perspective's C++ part with PGO would be a good idea.

I can suggest the following action points:
* Perform PGO benchmarks on Perspective. And if it shows improvements - add a note to the documentation about possible improvements in Perspective performance with PGO.
* Providing an easier way (e.g. a build option) to build scripts with PGO can be helpful for the end-users and maintainers since they will be able to optimize Perspective according to their own workloads.
* Optimize pre-built Perspective binaries with PGO.

Since the Perspective native part is the library, I think the Pydantic-core [experience](https://github.com/pydantic/pydantic-core/blob/main/Makefile#L66) can be reused here. Also, Clang [supports](https://clang.llvm.org/docs/UsersManual.html#profiling-with-instrumentation) PGO for shared libraries.

Maybe testing Post-Link Optimization techniques (like [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md)) would be interesting too (Clang and Rustc already use BOLT as an addition to PGO) but I recommend starting from the usual PGO.

Here are some examples of how PGO optimization is integrated in other projects:

* Rustc: a CI [script](https://github.com/rust-lang/rust/blob/master/src/ci/stage-build.py) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
  - A [part](https://github.com/gcc-mirror/gcc/blob/4832767db7897be6fb5cbc44f079482c90cb95a6/configure#L7818) in a "wonderful" `configure` script 
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html) 
* Python: 
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)
* file.d: [GitHub PR](https://github.com/ozontech/file.d/pull/469)
* OceanBase: [CMake flag](https://github.com/oceanbase/oceanbase/blob/master/cmake/Env.cmake#L55)

**Comment by texodus** [comment]
`perspective-python` benchmarks with PGO enabled (`master`) vs `v2.6.0`:
<img width="982" alt="Screenshot 2023-10-29 at 11 21 55 PM" src="https://github.com/finos/perspective/assets/60666/5b8dc950-3037-4a42-858f-a5b4827710b4">
It would be a bit a CI hassle to get this working, from my (limited) experience the profile data format is volatile and we already have multiple `llvm` toolchains in the mix.

**Reply by zamazan4ik** [reply]
Thanks a lot for the benchmark! Is it possible to perform "PGO vs non-PGO" benchmark on the same code version? Like "master PGO vs master non-PGO" or "2.6.0 PGO vs 2.6.0 non-PGO"? In this case, it would be much simpler to say that we got the performance boost just from PGO and not from the code difference between versions.

And what do we measure on the Y-axis? Time (in seconds/minutes/hours) or anything else?

**Reply by texodus** [reply]
There is no relevant code difference between these versions. You can learn more about our benchmark suite [here](https://github.com/finos/perspective/tree/master/tools/perspective-bench).

**Reply by zamazan4ik** [reply]
Thanks for the benchmark link!

> It would be a bit a CI hassle to get this working, from my (limited) experience the profile data format is volatile and we already have multiple llvm toolchains in the mix.

It could be the problem only if you decide to commit the profile to the repository and use it later (profile caching). You can try to implement the approach like it's done in other projects like [pydantic-code](https://github.com/pydantic/pydantic-core/blob/main/Makefile#L66) - generate profiles on-demand. So if you have multiple LLVM toolchains - you will need multiple on-demand PGO builds (for each LLVM toolchain). Which approach to implement - is up to you.
